Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,89 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a hook that allows managed code
+to run before the main application's entry point.
+
+## Motivation
+
+This would allow hosting providers to put configuration and policy in
+a single location, including settings that potentially influence load
+behavior of the main entry point such as the AssemblyLoadContext
+behavior. This could be used to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook needs to be separate from the
+entry point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+Environment variables or `<appname>.runtimeconfig.json` can be used to
+specify a managed assembly and type that contains an `Initialize`
+method.
+
+```
+DOTNET_MANAGED_HOST_ASSEMBLY=/path/to/ManagedHost.dll
+DOTNET_MANAGED_HOST_TYPE=ManagedHostNamespace.ManagedHostType
+```
+
+```
+{
+    "runtimeOptions": {
+        "managedHostAssembly": "/path/to/ManagedHost.dll",
```

Comment:
We may want to allow the case where the assembly is part of the app. 

It may be the best option for the service fabric case. Service fabric will drop a small ServiceFabricLoader.dll into each app and add it to .runtime.config.json.

This would save them from messing with environment variables, and also make the same app for across different machines.

Reply:
In general we'd like to avoid any injection / magic in the build process - so that the user code/package is just a regular package and they can use dotnet CLI commands to build it. Those package would have be launched via service fabric anyway so we have full control over the environment / lookup.

Reply:
But I do see an option that Service Fabric can inject this DLL at package upload time to become part of the app, but I don't see any significant difference between the two.

Reply:
> In general we'd like to avoid any injection / magic in the build process - so that the user code/package is just a regular package and they can use dotnet CLI commands to build it. 

I was not sure whether you prefer injecting magic at build time or injecting magic at launch time (you need to do one at least). Good to know that you prefer magic at launch time.

Reply:
I would like to consider the development process here as well - which should arguably be the most important consideration. What is the least confusing and unexpected for the developer of the app. Magic is fine for Azure, but during development time there should be 0 magic. Environment variables aren't always good if development is happening in VS - unless the project system can help with that and be reliable - I have serious doubts about this based on my experience working with the project system.

Reply:
Arguably, given that this is single-purpose (you can only set one managed host assembly), in practice this is always controlled by one single platform that is determined by the project / environment / application type, and therefore there are usually a "launcher" involved to launch them. I think this is a stop-gap measure specifically designed for these kind of scenarios.

In the future, one can imagine you can bring multiple of these host assemblies into the app, each determined from different nuget packages, and in that case I agree folks should never use environment variable in this case.

Reply:
@yizhang82 I am not sure if you are agreeing or not 😄 My point here is that during development users should not have a concern about their application being launched in a different way or unknown way that causes them to scramble to try and debug the scenario that is impossible for them to simulate locally. If your statement is 'the same host will be injected by the project system/template' so in theory this should be a no-op - yay! However, in my experience users try something locally, deploy and then assumptions about what will happen are broken and that makes me want to cry because it demonstrates a lack of developer focus and creates an environment for developers where they can't succeed.

Reply:
>> My point here is that during development users should not have a concern about their application being launched in a different way or unknown way that causes them to scramble to try and debug the scenario that is impossible for them to simulate locally

My point is that this is not true today for such platforms. Basically, by adding dependency to service fabric assemblies, you are effectively taking a dependency to the underlying service fabric platform and you cannot run it without service fabric environment - which is not "just load those assemblies".  Think of them as loosely coupled "plugins". Or imagine they are VS extension helper processes that can't run without VS and meanwhile need to load VS assemblies to be able to communicate with VS. 

I can see a scenario where a managed host assembly being injected by the project template, but I think it is important to keep the environment variable option for our production scenarios, and even for development scenarios (where we launch a mini-cluster on user machine which has the full environment). 

For cloud platforms, unfortunately, the biggest challenge is exactly you mentioned - how can you be sure that what you developed locally and make sure it works when deploy into the cloud where the assumptions are different, which is why it is important to do exactly what production does - spinning up a service fabric environment / cluster and launch the processes from there, even for development purposes. Even then it is still quite challenging, now that I'm on the other side of the fence here ;)

Reply:
> Basically, by adding dependency to service fabric assemblies, you are effectively taking a dependency to the underlying service fabric platform and you cannot run it without service fabric environment

If this is true, we are in bigger trouble than I thought. The whole point of having an IDE and projects is to have some reliability about debugging and designing a solution. I see how my concern is larger than the current review. Thanks for the unsettling information 😖